### ISSUE TITLE
Allow bootstrapping of Dask cluster on top of a PySpark cluster

### ISSUE THREAD

**Issue by holdenk** [dask/distributed#566]
For organization which have an existing PySpark installation being able to quickly bootstrap a distributed Dask installation could be useful and may be able to be a source of future integrations between PySpark and Dask.

Design on the Spark side wise we have two choices: running a task returning early and depending on the worker reuse code to keep our Python process alive, or similar to the Spark MR1 interface we could do an infinitely long map task and change the timeouts, or we could fork a process and hope it the container doesn't get cleaned up.

Explicitly out of scope:
Interaction with dynamic scheduling in Spark

Note:

Should https://issues.apache.org/jira/browse/SPARK-636 be implemented we could improve this implementation to not use a global lock to ensure only once per Spark python executor launching.

**Comment by mrocklin** [comment]
First failed attempt:

```python
def start_worker(host, port):
    from distributed import Worker
    from tornado.ioloop import IOLoop
    from threading import Thread
    import dask.distributed
    if hasattr(dask.distributed, 'global_worker'):  # lets be idempotent
       return
    
    loop = IOLoop.current()
    t = Thread(target=loop.start, daemon=True)
    t.start()
    w = Worker(host, port, loop=loop)
    w.start(0)
    dask.distributed.global_worker = w
    from time import sleep
    sleep(5)
    return [0]
```

**Comment by mrocklin** [comment]
Here is a less failed attempt that blocks the Python worker until the worker closes.  This blocks the thread until Dask work completes.

### Start a local dask scheduler on localhost (or in a local thread)

```
$ dask-scheduler
Scheduler started at localhost:8786
```

### Start a SparkContext and use it to launch Dask Workers

```python
from pyspark import SparkContext
sc = SparkContext('local[4]')

rdd = sc.parallelize(range(100))

def start_worker(host, port):
    import distributed
    from distributed import Worker
    from tornado.ioloop import IOLoop
    
    if getattr(distributed, 'global_worker', False):
       return ['already occupied']
    
    loop = IOLoop.current()
    w = Worker(host, port, loop=loop)
    w.start(0)

    distributed.global_worker = w
    
    from tornado import gen
    @gen.coroutine
    def _():
        while w.status != 'closed':
            yield gen.sleep(0.1)
    loop.run_sync(_)
    distributed.global_worker = False
    return ['completed']

rdd.mapPartitions(lambda x: start_worker('localhost', 8786)).count()
```

### In another process (or probably another thread)

```python
from dask.distributed import Client
c = Client('localhost:8786')
```

**Comment by mrocklin** [comment]
The start_worker function could be modified to take in the values for the current partition.  For example if we were to map this over a Spark DataFrame we would get a list of Row objects.  #567 

We can insert these directly into the worker.  We'll have to have the worker inform the scheduler that it has this data (this exists) and have the scheduler inform the relevant clients that this data exists and is something that they should care about (some sort of channel publication mechanism).  This doesn't exist, but has been requested a few times now.

**Comment by mrocklin** [comment]
@holdenk do you happen to have the incantation to control timeouts for long-running Python tasks?

**Comment by mrocklin** [comment]
@holdenk and I were looking at starting a long-lived Worker to run in the background from a short-lived PySpark task.  This would be good because it would release Spark resources and give us back control of the SparkContext thread.  Unfortunately it looks like PySpark cleaning up the Python processes pretty soon after the PySpark task completes, killing our worker-in-a-background-thread.  We tried to get past this by persisting a bit of data on the Spark side, but Spark still kills the Python process after about a minute, instead choosing to persist the data as serialized bytes on the Spark side.  

We could still get around this by spawning a separate process under the parent PySpark process that would live beyond the life of the forked PySpark child task process.

```
pyspark python parent process
  pyspark forked child task process
  pyspark forked child task process
    tried inserting a thread here, but this didn't work becasue the forked process gets cleaned up
  pyspark forked child task process
  start dask-worker process here <----
```

**Comment by mrocklin** [comment]
Another issue is that the worker needs a few open ports.  Spark is presumably deployed with something like Yarn that controls access to free ports.  It would be nice for workers to be able to identify a few free ports for them to use.

Some options:

1.  Use a random port (what we do by default)
2.  Choose a Spark Executor port number and add one (assuming that the Spark worker was handed a port range)
3.  Ask Yarn directly for a free port through some service

**Comment by mrocklin** [comment]
I threw the hacky implementation here up as a small package on PyPI.  I also added the reverse transformation of starting Spark from Dask.  This is now actually my default way of starting spark clusters now (just because I have lots of tooling to start dask clusters).

https://github.com/mrocklin/dask-spark

Thank you for your collaboration on this @holdenk .  I'm going to go ahead and close this issue for now.

**Comment by margitZ** [comment]
Hi - I'm trying to get the dask-spark code to work with a Spark + YARN cluster, but having problems starting the workers.  Is it possible to adapt dask-spark to work with dask_yarn?

**Comment by martindurant** [comment]
The approach here (starting dask workers within containers managed by spark) is very different from the one in dask_yarn (interact with yarn directly to launch python processes which run dask). I'm not sure I see a way that the two could be used together. If you have tried dask-yarn and had problems, you may want to report them on the issue tracker of that project.
